File: backend/app/schemas/AuditEvent.py
from pydantic import BaseModel, field_validator
from datetime import datetime
from typing import Optional, Dict, Any
from app.schemas.user import UserOut
import json

class AuditEventBase(BaseModel):
    event_type: str
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    details: Optional[Dict[str, Any]] = None

class AuditEventCreate(AuditEventBase):
    user_id: Optional[int] = None
    target_user_id: Optional[int] = None

class AuditEventOut(AuditEventBase):
    id: int
    user_id: Optional[int]
    target_user_id: Optional[int]
    created_at: datetime
    user: Optional[UserOut] = None
    target_user: Optional[UserOut] = None

    class Config:
        from_attributes = True

    @field_validator('details', mode='before')
    @classmethod
    def parse_details(cls, v):
        """Parse JSON string to dict if needed"""
        if v is None:
            return None
        if isinstance(v, str):
            try:
                return json.loads(v)
            except json.JSONDecodeError:
                return {"raw": v}  # Fallback si le JSON est invalide
        if isinstance(v, dict):
            return v
        return {"value": str(v)}  # Fallback pour autres types

class AuditEventFilter(BaseModel):
    event_type: Optional[str] = None
    user_id: Optional[int] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    limit: int = 100
    offset: int = 0

    class Config:
        from_attributes = True

File: shareholder-dashboard/src/core/types/AuditEvent.ts
import { User } from "./type";

export interface AuditEvent {
    id: number;
    event_type: string;
    user_id?: number;
    target_user_id?: number;
    ip_address?: string;
    user_agent?: string;
    details?: any;
    created_at: string;
    user?: User;
    target_user?: User;
}

export interface AuditEventFilter {
    event_type?: string;
    user_id?: number;
    start_date?: string;
    end_date?: string;
    limit?: number;
    offset?: number;
}

export interface AuditStats {
    total_events: number;
    today_events: Record<string, number>;
    active_users_today: number;
}